common/dynamicconfig/collection.go

```diff
@@ -0,0 +1,543 @@
+	return func(namespaceID string) bool {
+		return matchAndConvert(
+			c,
+			key,
+			defaultValue,
+			namespaceIDPrecedence(namespaceID),
+			convertBool,
+		)
+	}
+}
+
+// GetBoolPropertyFilteredByTaskQueueInfo gets property with taskQueueInfo as filters and asserts that it's a bool
+func (c *Collection) GetBoolPropertyFilteredByTaskQueueInfo(key Key, defaultValue any) BoolPropertyFnWithTaskQueueInfoFilters {
+	return func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) bool {
+		return matchAndConvert(
+			c,
+			key,
+			defaultValue,
+			taskQueuePrecedence(namespace, taskQueue, taskType),
+			convertBool,
+		)
+	}
+}
+
+// Task queue partitions use a dedicated function to handle defaults.
+func (c *Collection) GetTaskQueuePartitionsProperty(key Key) IntPropertyFnWithTaskQueueInfoFilters {
+	return c.GetIntPropertyFilteredByTaskQueueInfo(key, defaultNumTaskQueuePartitions)
+}
+
+func findMatch(cvs, defaultCVs []ConstrainedValue, precedence []Constraints) (any, error) {
+	if len(cvs)+len(defaultCVs) == 0 {
+		return nil, errKeyNotPresent
+	}
+	for _, m := range precedence {
+		// duplicate the code so that we don't have to allocate a new slice to hold the
+		// concatenation of cvs and defaultCVs
+		for _, cv := range cvs {
+			if m == cv.Constraints {
+				return cv.Value, nil
+			}
+		}
+		for _, cv := range defaultCVs {
+			if m == cv.Constraints {
+				return cv.Value, nil
+			}
+		}
+	}
+	// key is present but no constraint section matches
+	return nil, errNoMatchingConstraint
+}
+
+// matchAndConvert can't be a method of Collection because methods can't be generic, but we can
+// take a *Collection as an argument.
+func matchAndConvert[T any](
+	c *Collection,
+	key Key,
+	defaultValue any,
+	precedence []Constraints,
+	converter func(value any) (T, error),
+) T {
+	cvs := c.source.GetValue(key)
+
+	// defaultValue may be a list of constrained values. In that case, one of them must have an
+	// empty constraint set to be the fallback default. Otherwise we'll return the zero value
+	// and log an error (since []ConstrainedValue can't be converted to the desired type).
+	defaultCVs, _ := defaultValue.([]ConstrainedValue)
+
+	val, matchErr := findMatch(cvs, defaultCVs, precedence)
+	if matchErr != nil {
+		c.logError(key, matchErr)
+		// couldn't find a constrained match, use default
+		val = defaultValue
+	}
+
+	typedVal, convertErr := converter(val)
+	if convertErr != nil && matchErr == nil {
+		// We failed to convert the value to the desired type. Try converting the default. note
+		// that if err != nil then val _is_ defaultValue and we don't have to try this again.
+		c.logError(key, convertErr)
+		typedVal, convertErr = converter(defaultValue)
+	}
+	if convertErr != nil {
+		// If we can't convert the default, that's a bug in our code, use Warn level.
+		c.logger.Warn("can't convert default value", tag.Key(key.String()), tag.Error(matchErr))
```

Comment:
add message to indicate this his a bug that needs attention. For example: This is a bug, fix me.